Add RPC chapter (#30)
tikv/deep-dive-tikv: src/rpc/grpc.md

# gRPC

> Sitting on top of HTTP/2 and protobuf gRPC provides a high performance RPC

Reviewer: ```suggestion
Sitting on top of HTTP/2 and protobuf, gRPC provides a high-performance RPC
```

Reviewer 2: ```suggestion
Sitting on top of HTTP/2 and protocol buffers gRPC provides a high performance RPC
```

I would also add links for http/2 and protobufs.

Should we summarise what RPC is?

Reviewer 2: Ah, I see there is an RPC intro before this, ignore the second comment

framework. It supports features such as bi-directional streaming and authentication.

gRPC is also part of the [CNCF](https://www.cncf.io/projects/). It is commonly
used in applications like Kubernetes or TiKV.

## Servers and Stubs

gRPC works off the idea that there is a *server* node which can accept

> *requests* from *stub* client nodes.

Reviewer 2: I think it is actual clients which send request, not client stubs, the stubs are generated by the gRPC library and then filled out by the programmer, but the stubs alone don't do much.

The *server* implements the service interface and runs a gRPC server to handle

> all the incoming RPC calls from the client *stubs*. The *stubs* which clients

Reviewer: ```suggestion
all the incoming RPC calls from the client *stubs*. The client *stubs*
```

> all the incoming RPC calls from the client *stubs*. The *stubs* which clients

Reviewer: ```suggestion
all the incoming RPC calls from the client *stubs*. The client *stubs*
```

> have provide the same methods that the server implements, but instead send the

Reviewer: ```suggestion
provide the same methods that the server implements, but instead send the
```

request to the server.

![GRPC Server and Stub connections](grpc-server-stub.svg)

## Streams

Sometimes it's convienent to return or accept a stream of messages, instead of
operating in the traditional 1:1 request response format.

> gRPC supports both of these, enabling bi-directional streaming so clients and

Reviewer 2: what does "both of these" mean? 1:1 and streaming or client streaming and server streaming?

servers can write in whatever order they like. gRPC maintains ordering behind
the scenes.

```protobuf
service StreamScan {
  rpc BatchGet (string, string) returns (stream string) {};
}
```

## A Rust Crate

One of the TiKV maintaining organizations, PingCAP, maintains the
[`grpc-rs`](https://github.com/pingcap/grpc-rs) library, providing Rust bindings

> to the C library of gRPC.

Reviewer 2: Given there are other grpc crates, it might be good to say that TiKV uses grpc-rs
